File: lab1.py
import os
import sys

def main():
    try:
        # 1. Create a child process
        # This maps to: int pid; pid = fork();
        pid = os.fork()
    except OSError:
        # This maps to: if (pid < 0)
        print("fork failed")
        sys.exit(1)

    if pid == 0:
        # --- CHILD PROCESS ---
        # This maps to: else if (pid == 0)
        
        # execlp(file, arg0, arg1, ...)
        # The C code used "1s" as the second argument (argv[0]), so we match that here.
        # This replaces the current process image with 'whoami'.
        try:
            os.execlp("whoami", "1s")
        except OSError as e:
            print(f"Execution failed: {e}")
            sys.exit(0)
            
    else:
        # --- PARENT PROCESS ---
        # This maps to: else
        
        # print the current (Parent) Process ID
        print(f"\n Process ID is: {os.getpid()}")
        
        # Wait for the child process to finish
        # This maps to: wait(NULL);
        os.wait()
        
        sys.exit(0)

if __name__ == "__main__":
    main()

File: lab1.cs
using System;
using System.Diagnostics;

namespace SystemCallsLab
{
    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                // In C code: printf("\n Process ID is: %d\n", getpid());
                // We print the Parent's PID first
                int parentPid = Process.GetCurrentProcess().Id;
                Console.WriteLine($"\n Process ID is: {parentPid}");

                // In C code: execlp("whoami", "1s", NULL);
                // We configure the child process to run "whoami"
                ProcessStartInfo startInfo = new ProcessStartInfo();
                startInfo.FileName = "whoami";
                startInfo.UseShellExecute = false; // Required to not use the OS shell
                
                // create the child process
                using (Process childProcess = Process.Start(startInfo))
                {
                    // In C code: wait(NULL);
                    // The parent waits for the child to finish
                    childProcess.WaitForExit();
                }
            }
            catch (Exception ex)
            {
                // In C code: printf("fork failed");
                Console.WriteLine($"Process creation failed: {ex.Message}");
                Environment.Exit(1);
            }

            Environment.Exit(0);
        }
    }
}